File: sdk/node/hpkv-rioc/examples/mtls.ts
import { RiocClient, RiocConfig } from '../src';

async function main() {
  // Create client configuration with mTLS
  const config: RiocConfig = {
    host: '127.0.0.1',
    port: 8000,
    timeoutMs: 5000,
    tls: {
      caPath: '/workspaces/kernel-high-performance-kv-store/api/rioc/certs/ca.crt',
      certificatePath: '/workspaces/kernel-high-performance-kv-store/api/rioc/certs/client.crt',
      keyPath: '/workspaces/kernel-high-performance-kv-store/api/rioc/certs/client.key',
      verifyHostname: '127.0.0.1',
      verifyPeer: true
    }
  };

  console.log('Client Configuration:');
  console.log(`  Host:      ${config.host}`);
  console.log(`  Port:      ${config.port}`);
  console.log(`  CA Path:   ${config.tls?.caPath}`);
  console.log(`  Cert Path: ${config.tls?.certificatePath}`);
  console.log(`  Key Path:  ${config.tls?.keyPath}`);
  console.log('\nConnecting to server...');

  const startTime = process.hrtime.bigint();

  // Create client
  const client = new RiocClient(config);
  const connectTime = Number(process.hrtime.bigint() - startTime) / 1_000_000;
  console.log(`Connected in ${connectTime.toFixed(2)} ms`);

  try {
    // Test data
    const key = Buffer.from('mtls_test_key');
    const value = Buffer.from('mtls test value');

    // Insert
    const timestamp = RiocClient.getTimestamp();
    console.log(`\n1. Inserting record with timestamp ${timestamp}`);
    const insertStart = process.hrtime.bigint();
    client.insert(key, value, timestamp);
    const insertTime = Number(process.hrtime.bigint() - insertStart) / 1_000_000;
    console.log(`Insert successful in ${insertTime.toFixed(2)} ms`);

    // Get
    console.log('\n2. Getting record');
    const getStart = process.hrtime.bigint();
    const retrievedValue = client.get(key);
    const getTime = Number(process.hrtime.bigint() - getStart) / 1_000_000;
    if (retrievedValue === null) {
      console.log(`Key not found (took ${getTime.toFixed(2)} ms)`);
    } else {
      console.log(`Get successful in ${getTime.toFixed(2)} ms, value: ${retrievedValue.toString()}`);
    }

    // Delete
    const deleteTimestamp = RiocClient.getTimestamp();
    console.log('\n3. Deleting record');
    const deleteStart = process.hrtime.bigint();
    client.delete(key, deleteTimestamp);
    const deleteTime = Number(process.hrtime.bigint() - deleteStart) / 1_000_000;
    console.log(`Delete successful in ${deleteTime.toFixed(2)} ms`);

  } finally {
    client.dispose();
    console.log('\nTest completed');
  }
}

main().catch(error => {
  console.error('Error:', error);
  process.exit(1);
}); 

File: sdk/python/hpkv-rioc/examples/mtls.py
"""
Example of using mTLS with the HPKV RIOC Python SDK.
"""

import argparse
import os
from pathlib import Path

from hpkv_rioc import RiocClient, RiocConfig, RiocTlsConfig

def get_default_cert_paths():
    """Get default certificate paths from the RIOC certs directory."""
    workspace_root = ""
    certs_dir = os.path.join(workspace_root, "")
    return {
        "client_cert": os.path.join(certs_dir, "client.crt"),
        "client_key": os.path.join(certs_dir, "client.key"),
        "ca_cert": os.path.join(certs_dir, "ca.crt")
    }

def main():
    # Parse command line arguments
    parser = argparse.ArgumentParser(description="RIOC mTLS Example")
    cert_paths = get_default_cert_paths()
    
    parser.add_argument("--client-cert", default=cert_paths["client_cert"],
                      help="Path to client certificate file")
    parser.add_argument("--client-key", default=cert_paths["client_key"],
                      help="Path to client private key file")
    parser.add_argument("--ca-cert", default=cert_paths["ca_cert"],
                      help="Path to CA certificate file")
    parser.add_argument("--host", default="localhost",
                      help="Server hostname")
    parser.add_argument("--port", type=int, default=8000,
                      help="Server port")
    
    args = parser.parse_args()

    # Verify certificate files exist
    for cert_path in [args.client_cert, args.client_key, args.ca_cert]:
        if not os.path.exists(cert_path):
            print(f"Error: Certificate file not found: {cert_path}")
            return

    # Create TLS configuration
    tls_config = RiocTlsConfig(
        certificate_path=args.client_cert,
        key_path=args.client_key,
        ca_path=args.ca_cert,
        verify_hostname=args.host,
        verify_peer=True
    )

    # Create client configuration with TLS
    config = RiocConfig(
        host=args.host,
        port=args.port,
        timeout_ms=5000,
        tls=tls_config
    )

    print(f"Connecting to {args.host}:{args.port} with mTLS...")
    print(f"Using certificates:")
    print(f"  Client cert: {args.client_cert}")
    print(f"  Client key:  {args.client_key}")
    print(f"  CA cert:     {args.ca_cert}")

    # Create client
    client = RiocClient(config)

    # Insert some key-value pairs
    print("\nInserting key-value pair...")
    client.insert_string("secure-key", "secure-value")

    # Get value
    print("\nRetrieving value...")
    value = client.get_string("secure-key")
    print(f"secure-key -> {value}")  # secure-key -> secure-value

    # Use batch operations with TLS
    print("\nPerforming batch operations...")
    with client.batch() as batch:
        batch.add_insert(b"key1", b"value1", client.get_timestamp())
        batch.add_get(b"key1")

    print("\nOperations completed successfully!")

if __name__ == "__main__":
    main() 